ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -0,0 +1,264 @@
+#include "../PrjFSKext/VirtualizationRoots.hpp"
+#include "../PrjFSKext/VirtualizationRootsPrivate.hpp"
+#include "../PrjFSKext/PrjFSProviderUserClient.hpp"
+#include "../PrjFSKext/kernel-header-wrappers/mount.h"
+#import <XCTest/XCTest.h>
+#include "../PrjFSKext/Memory.hpp"
+#include "../PrjFSKext/Locks.hpp"
+#include "../PrjFSKext/VnodeUtilities.hpp"
+#include "../PrjFSKext/KextLog.hpp"
+#include "KextMockUtilities.hpp"
+#include "MockVnodeAndMount.hpp"
+#include <vector>
+#include <string>
+
+class PrjFSProviderUserClient
+{
+};
+
+
+void ProviderUserClient_UpdatePathProperty(PrjFSProviderUserClient* userClient, const char* providerPath)
+{
+    MockCalls::RecordFunctionCall(ProviderUserClient_UpdatePathProperty, userClient, providerPath);
+}
+
+
+@interface VirtualizationRootsTests : XCTestCase
+
+@end
+
+@implementation VirtualizationRootsTests
+{
+    PrjFSProviderUserClient dummyClient;
+    pid_t dummyClientPid;
+    MountPointer testMountPoint;
+}
+
+- (void)setUp
+{
+    srand(0);
+    self->dummyClientPid = 100;
+    
+    // This is roughly what "real" fsids look like
+    fsid_t testMountFsid = { (1 << 24) | (rand() % 32), (rand() % 16) };
+    
+    // HFS+ inodes are < UINT32_MAX (APFS' are 64-bit)
+    uint64_t testMountInitialInode = rand();
+    
+    testMountPoint = mount::Create("hfs", testMountFsid, testMountInitialInode);
+    
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+}
+
+- (void)tearDown
+{
+    VirtualizationRoots_Cleanup();
+
+    MockVnodes_CheckAndClear();
+    MockCalls::Clear();
+}
+
+
+- (void)testRegisterProviderForPath_NonexistentPath
+{
+    const char* path = "/Users/test/code/Does/Not/Exist";
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, ENOENT);
+    XCTAssertFalse(VirtualizationRoot_IsValidRootHandle(result.root));
+    
+    XCTAssertFalse(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+}
+
+- (void)testRegisterProviderForPath_NonDirectoryPath
+{
+    const char* path = "/Users/test/code/NotADirectory.cpp";
+    
+    VnodePointer vnode = vnode::Create(self->testMountPoint, path);
+    
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, ENOTDIR);
+    XCTAssertFalse(VirtualizationRoot_IsValidRootHandle(result.root));
+
+    XCTAssertFalse(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+}
+
+- (void)testRegisterProviderForPath_DisallowedFileSystem
+{
+    const char* path = "/Volumes/USBStick/repo";
+    
+    fsid_t fatFsid = self->testMountPoint->GetFsid();
+    fatFsid.val[1]++;
+    MountPointer fatMount(mount::Create("msdos", fatFsid, rand()));
```

Comment:
arguably this could remove the need to test VirtualizationRoot_VnodeIsOnAllowedFilesystem directly.

If we want to do that we'd need a test for both "hfs" and "apfs".

Reply:
I think we should definitely keep the unit tests for `VirtualizationRoot_VnodeIsOnAllowedFilesystem`. If anything, we should probably allow swapping out the `VirtualizationRoot_VnodeIsOnAllowedFilesystem` function itself for a test double to make this test here simpler.

Reply:
There is a train of though on unit tests that private functions should never be tested (not part of the interface).  Of course for the kext with the majority being private that really doesn't hold well.  

I'd be careful with mocking out`VirtualizationRoot_VnodeIsOnAllowedFilesystem`.  While it's true we actually testing two functions at the same time, it's also how it's used.  I've got mixed feelings there.  Once we get more converge this would be great to discuss in more detail :-)!

Reply:
For `VirtualizationRoot_VnodeIsOnAllowedFilesystem` it probably doesn't matter too much, but I think we'll need to be able to swap out some of the other functions, as always testing the whole call subtree via its side effects is going to be awkward.

The public/private distinction isn't terribly helpful in our kext; the only truly public APIs are the kext start/stop functions, the callbacks we register with the kauth listener, and the IOKit userclient classes. We're also treating certain compilation units as class-like with APIs that are visible to the rest of the kext vs local static functions, but I think those only superficially map to class-based OOP design; we'd probably have more, smaller classes in practice if we were writing this in C# or Java or whatever.